### PR TITLE
[BUG] Remove empty tag values

### PR DESCRIPTION
**Pull request by ashtul**
The fix updates ForkGC to remove empty tag values from the trie.
This bug affected indexes where TAG fields are updated with new values while the old, deleted values are left in the trie.

### CODE REVIEW

File: src/fork_gc.c [hunk 1]
```diff
@@ -1006,11 +1021,18 @@ static FGCError FGC_parentHandleTags(ForkGC *gc, RedisModuleCtx *rctx) {
...
+    // No more tags values in tag field
     if (value == NULL) {
       RS_LOG_ASSERT(status == FGC_COLLECTED, "GC status is COLLECTED");
       break;
     }
 
+    char *tagVal;
+    size_t tagValLen;
+    if (FGC_recvBuffer(gc, (void **)&tagVal, &tagValLen) != REDISMODULE_OK) {
+      return FGC_PARENT_ERROR;
```
Comment (MeirShpilraien) [inline src/fork_gc.c:1033]: Why is this a parrent error? also if you return here you do not free `header.field` no?

Comment (MeirShpilraien) [reply src/fork_gc.c:1033]: Also, it's better to follow the other variables and put declaration on loop start and initialize them.

File: src/fork_gc.c [hunk 2]
```diff
@@ -368,22 +372,25 @@ static void sendNumericTagHeader(ForkGC *fgc, void *arg) {
...
+    FGC_sendBuffer(fgc, info->tagValue, strlen(info->tagValue));
+  }
 }
 
 // If anything other than FGC_COLLECTED is returned, it is an error or done
-static FGCError recvNumericTagHeader(ForkGC *fgc, char **fieldName, size_t *fieldNameLen,
-                                     uint64_t *id) {
-  if (FGC_recvBuffer(fgc, (void **)fieldName, fieldNameLen) != REDISMODULE_OK) {
+static FGCError recvNumericTagHeader(ForkGC *fgc, tagNumHeader *header, int type) {
+  size_t tagValLen;
```
Comment (MeirShpilraien) [inline src/fork_gc.c:382]: Do you use this?

File: src/fork_gc.c [hunk 3]
```diff
@@ -368,22 +372,25 @@ static void sendNumericTagHeader(ForkGC *fgc, void *arg) {
...
+  if (info->type == RSFLDTYPE_TAG) {
+    FGC_sendBuffer(fgc, info->tagValue, strlen(info->tagValue));
+  }
 }
 
 // If anything other than FGC_COLLECTED is returned, it is an error or done
-static FGCError recvNumericTagHeader(ForkGC *fgc, char **fieldName, size_t *fieldNameLen,
-                                     uint64_t *id) {
-  if (FGC_recvBuffer(fgc, (void **)fieldName, fieldNameLen) != REDISMODULE_OK) {
+static FGCError recvNumericTagHeader(ForkGC *fgc, tagNumHeader *header, int type) {
```
Comment (MeirShpilraien) [inline src/fork_gc.c:381]: Where do you use the type argument?

File: src/fork_gc.c [hunk 4]
```diff
@@ -899,7 +911,7 @@ static void resetCardinality(NumGcInfo *info, NumericRangeNode *currNone) {
   r->card = array_len(r->values);
 }
 
-static void applyNumIdx(ForkGC *gc, RedisSearchCtx *sctx, NumGcInfo *ninfo) {
+static void applyNumIdx(ForkGC *gc, RedisSearchCtx *sctx, NumGcInfo *ninfo, NumericRangeTree *rt) {
```
Comment (MeirShpilraien) [inline src/fork_gc.c:914]: Let not change the signiture here until we know we need it for the numeric fix.

File: tests/pytests/test_tags.py [hunk 5]
```diff
@@ -193,3 +193,50 @@ def testTagCaseSensitive(env):
...
+    env.expect('DEL doc1').equal(1)
+    env.expect('DEL doc2').equal(1)
+
+    forceInvokeGC(env, 'idx')
+
+    # make sure the inverted index was cleaned
+    env.expect('FT.DEBUG DUMP_TAGIDX idx t').equal([])
+
+    # read from the cursor
+    res, cursor = env.cmd('FT.CURSOR READ idx %d' % cursor)
```
Comment (MeirShpilraien) [inline tests/pytests/test_tags.py:239]: Lets also add this same test but put some new data after the term was deleted

**Review by MeirShpilraien — COMMENTED** [review]
:+1: 
Some small comments

**Review by ashtul — COMMENTED** [review]
@MeirShpilraien made requested changes

File: src/fork_gc.c [hunk 6]
```diff
@@ -1000,17 +1015,25 @@ static FGCError FGC_parentHandleTags(ForkGC *gc, RedisModuleCtx *rctx) {
     MSG_IndexInfo info = {0};
     InvIdxBuffers idxbufs = {0};
     TagIndex *tagIdx = NULL;
+    char *tagVal;
```
Comment (MeirShpilraien) [inline src/fork_gc.c:1018]: Initialize to NULL

File: src/tag_index.c [hunk 7]
```diff
@@ -134,8 +136,8 @@ size_t TagIndex_Index(TagIndex *idx, const char **values, size_t n, t_docId docI
 }
 
 typedef struct {
+  TagIndex *idx;
   IndexIterator **its;
-  uint32_t uid;
```
Comment (MeirShpilraien) [inline src/tag_index.c:138]: We do not use this?

File: tests/pytests/test_tags.py [hunk 8]
```diff
@@ -193,3 +193,96 @@ def testTagCaseSensitive(env):
...
+    env.expect('DEL', 'doc1').equal(1)
+    env.expect('DEL', 'doc2').equal(1)
+
+    forceInvokeGC(env, 'idx')
+
+    # make sure the inverted index was cleaned
+    env.expect('FT.DEBUG', 'DUMP_TAGIDX', 'idx', 't').equal([])
+
+    # add data
+    conn.execute_command('HSET', 'doc3', 't', 'foo')
```
Comment (MeirShpilraien) [inline tests/pytests/test_tags.py:277]: :+1:

**Review by MeirShpilraien — COMMENTED** [review]
Last comment

File: src/tag_index.c (hunk 7 above)
Comment (ashtul) [reply src/tag_index.c:138]: @MeirShpilraien 
Since I use `TagIndex` with `TagIndex_OpenIndex` to retrieve the inverted index of the tag value.

**Review by MeirShpilraien — APPROVED** [review]

**Pull request merged**